File: server/samples/clients/node/appencryption_client.js
/*
Asherah Server gRPC - Node client

A simple Node application that demonstrates integrating with Asherah Server
via a dynamically generated gRPC client.
*/

'use strict';

const yargs = require('yargs');
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const sleep = require('sleep');
const path = require('path');

// Setup logger
const winston = require('winston');
const consoleTransport = new winston.transports.Console();
const myWinstonOptions = {
  transports: [consoleTransport],
  format:
        winston.format.combine(
          winston.format.colorize({ all: true, }),
          winston.format.simple()
        ),
};
const logger = new winston.createLogger(myWinstonOptions);

let socketFile = '';

/**
 * SessionClient provides a synchronous client interface for the bidirectionally-streaming Session
 endpoint.
 */
class SessionClient {
    #call;
    #getSessionResolve;
    #encryptResolve;
    #decryptResolve;

    constructor(appEncryptionProto) {
      let drr;
      let bytes;
      let decryptedPayload;

      const client = new appEncryptionProto.asherah.apps.server.AppEncryption(`unix://${socketFile}`,
        grpc.credentials.createInsecure());

      const self = this;

      const call = client.Session();
      call.on('end', function () {
        logger.info('end of transmission from server');
      });
      call.on('error', function (err) {
        logger.error(err);
      });
      call.on('status', function (status) {
        logger.info(status);
      });
      call.on('data', function (sessionResponse) {
        switch (sessionResponse.response) {
          case 'encrypt_response':
            drr = sessionResponse.encrypt_response.data_row_record;
            logger.info('received DRR');
            self.#encryptResolve(drr);
            break;
          case 'decrypt_response':
            bytes = sessionResponse.decrypt_response.data;
            decryptedPayload = Buffer.from(bytes).toString('utf-8');
            logger.info(`received decrypted data: ${decryptedPayload}`);
            self.#decryptResolve(decryptedPayload);
            break;
          case 'error_response':
            logger.info('error received: ' + sessionResponse.error_response.message);
            break;
          default:
            self.#getSessionResolve();
        }
      });

      this.#call = call;
    }

    /**
     * Create a new session for communicating with the gRPC server
     * @param {string} partition - The partition id for the session
     * @returns {Promise<void>} - A session_response wrapped in a promise
     */
    getSession(partition) {
      const self = this;
      return new Promise(resolve => {
        self.#getSessionResolve = resolve;

        self.#call.write({ get_session: { partition_id: partition, }, });
      });
    }

    /**
     * Encrypt data using the current session and its partition.
     * @param {string} payload - The payload to be encrypted
     * @returns {Promise<void>} - An encrypt_response wrapped in a promise
     */
    encrypt(payload) {
      const self = this;
      return new Promise(resolve => {
        self.#encryptResolve = resolve;
        self.#call.write({ encrypt: { data: Buffer.from(payload), }, });
      });
    }

    /**
     * Decrypt the data using the current session and its partition
     * @param {bytes[]} drr - An encrypted data row record
     * @returns {Promise<void>} - A decrypt_response wrapped in a promise
     */
    decrypt(drr) {
      const self = this;
      return new Promise(resolve => {
        self.#decryptResolve = resolve;
        // eslint-disable-next-line camelcase
        self.#call.write({ decrypt: { data_row_record: drr, }, });
      });
    }

    /**
     * Close the stream
     */
    close() {
      this.#call.end();
    }
}

/**
 * Generate a random string of fixed length
 * @param {number} length - Length of string to be generated
 * @returns {string} - A random string
 */
function randomString(length = 12) {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }

  return result;
}

/**
 * Runs the client
 * @param {SessionClient} client - A session client for the current session
 * @returns {Promise<void>} - Client response wrapped in a promise
 */
async function runClient(client) {
  const payload = randomString();

  logger.info(`encrypting payload ${payload}`);
  const drr = await client.encrypt(payload);

  logger.info(`calling decrypt`);
  const decryptedPayload = await client.decrypt(drr);

  if (decryptedPayload !== payload) {
    logger.info('oh no... something went terribly wrong!');
  } else {
    logger.info(`test completed successfully\n`);
  }
}

/**
 * Executes runClient once
 * @param {SessionClient} client - A session client for the current session
 * @returns {Promise<void>} - Client response wrapped in a promise
 */
async function runOnce(client) {
  await runClient(client);

  client.close();
}

/**
 * Executes runClient until the process is interrupted.
 * @param {SessionClient} client - A session client for the current session
 * @returns {Promise<void>} - Client response wrapped in a promise
 */
async function runContinuously(client) {
  // eslint-disable-next-line no-constant-condition
  while (true) {
    process.on('SIGINT', () => {
      logger.info('received SIGINT. terminating client');
      process.exit(0);
    });
    process.on('SIGTERM', () => {
      logger.info('received SIGTERM. terminating client');
      process.exit(0);
    });
    await runClient(client);

    //  Sleep for 1 second before sending the next request
    sleep.sleep(1);
  }
}

/**
 * Gets a session and calls run_once or run_continuously based on the value of the continuous parameter
 * @param {boolean} continuous - Flag indicating if client should run continuously or not
 * @param {string} id - Unique identifier for the partition
 * @param {GrpcObject} appEncryptionProto - A gRPC object
 * @returns {Promise<void>} - Client response wrapped in a promise
 */
async function run(continuous, id, appEncryptionProto) {
  const client = new SessionClient(appEncryptionProto);

  await client.getSession(`partition-${id}`);

  if (continuous) {
    await runContinuously(client);
  } else {
    await runOnce(client);
  }
}

/**
 * Dynamically load the .proto file with the specified options and subsequently the gRPC package definition
 * @param {string} protoPath - Path of the proto file
 * @returns {GrpcObject} - A gRPC object
 */
function loadProtoDef(protoPath) {
  const appEncryptionDef = protoLoader.loadSync(protoPath, {
    keepCase: true,
    defaults: true,
    oneofs: true,
  });

  const appEncryptionProto = grpc.loadPackageDefinition(appEncryptionDef);

  return appEncryptionProto;
}

function main() {
  const requests = [];
  const argv = yargs.usage('Usage: $0 --socket [string] --continuous [boolean] ' +
        '--num-clients [num] --proto-path [string]')
    .options({
      s: {
        alias: 'socket',
        describe: 'The unix domain socket the server is listening on.',
        type: 'string',
        demand: false,
        default: '/tmp/appencryption.sock',
      },
      c: {
        alias: 'continuous',
        describe: 'When present the client will run a continuous test.',
        type: 'boolean',
        demand: false,
        default: false,
      },
      n: {
        alias: 'num-clients',
        describe: 'The total number of clients to run asynchronously',
        type: 'num',
        demand: false,
        default: 1,
      },
      p: {
        alias: 'proto-path',
        describe: 'The path to the proto file for the service',
        type: 'string',
        demand: false,
        default: path.join(__dirname, '../../../protos/appencryption.proto'),
      },
    })
    .argv;

  socketFile = argv.s;
  logger.info('starting test');

  const appEncryptionProto = loadProtoDef(path.resolve(argv.p));

  for (let i = 1; i <= argv.n; i++) {
    const promise = new Promise(() => run(argv.c, i, appEncryptionProto, () => {
      logger.info('client terminated');
    })
    );
    requests.push(promise);
  }

  Promise.all(requests);
}

if (require.main === module) {
  main();
}

File: server/samples/clients/python/appencryption_client.py
# -*- coding: utf-8 -*-

"""Asherah Server gRPC - Python client

A simple Python application that demonstrates integrating with Asherah Server
via a generated gRPC client.
"""

import argparse
import asyncio
import logging
import queue
import random
import signal
import string
import sys
from types import FrameType, TracebackType
from typing import Any, Iterator, Optional, Type

import grpc
import appencryption_pb2
import appencryption_pb2_grpc
from appencryption_types import SessionRequest, SessionResponse


class SessionReceiveError(Exception):
    """Raised when a gRPC error message is received by the client."""


class SessionClient:
    """A synchronous Session wrapper

    SessionClient provides a synchronous client interface for the bidirectionally-streaming Session
    endpoint.
    """
    requests: 'queue.Queue[SessionRequest]' = queue.Queue()
    session: Optional[Iterator]

    def __init__(self, socket: str, partition: str) -> None:
        self.requests = queue.Queue()

        self.channel = grpc.insecure_channel(f'unix://{socket}')
        self.stub = appencryption_pb2_grpc.AppEncryptionStub(self.channel)
        self.partition = partition

        self.session = None

    def __enter__(self) -> 'SessionClient':
        self._enter()
        return self

    def __exit__(
            self,
            exc_type: Optional[Type[BaseException]],
            exc_value: Optional[BaseException],
            traceback: Optional[TracebackType]
    ) -> None:
        self._close()

    def _close(self) -> None:
        self.channel.close()

    def _enter(self) -> None:
        req = appencryption_pb2.SessionRequest()
        req.get_session.partition_id = self.partition

        self._send_receive(req)

    def _send_receive(
            self,
            req: SessionRequest,
    ) -> SessionResponse:
        self.requests.put(req)

        if self.session is None:
            self.session = self.stub.Session(self._next_request())

        resp = next(self.session)
        if resp.HasField('error_response'):
            raise SessionReceiveError(resp.error_response.message)

        return resp

    def _next_request(self) -> Iterator[SessionRequest]:
        while True:
            yield self.requests.get()

    def encrypt(self, data: bytes) -> Any:
        """Encrypt data using the current session and its partition.

        Args:
            data: A bytes object containing the data to be encrypted.
        Returns:
            An appencryption_pb2.DataRowRecord containing the encrypted data,
            as well as its encrypted key and metadata.
        """
        req = appencryption_pb2.SessionRequest()
        req.encrypt.data = data

        resp = self._send_receive(req)

        return resp.encrypt_response.data_row_record

    def decrypt(self, drr: Any) -> bytes:
        """Decrypt the data using the current session and its partition.

        Args:
            drr: An appencryption_pb2.DataRowRecord containing the data to be decrypted.
        Returns:
            The decrypted data as bytes.
        """
        req = appencryption_pb2.SessionRequest()
        req.decrypt.data_row_record.CopyFrom(drr)

        resp = self._send_receive(req)

        return resp.decrypt_response.data


class InterruptHandler:
    """A simple signal handler."""

    def __init__(self) -> None:
        self.interrupted = False

    # pylint: disable=unused-argument
    def _interrupt(self, sig: int, frame: FrameType):
        logging.info('received signal %s', signal.Signals(sig).name)
        self.interrupted = True

    def start(self) -> None:
        """Start handling signals."""
        signal.signal(signal.SIGINT, self._interrupt)
        signal.signal(signal.SIGQUIT, self._interrupt)
        signal.signal(signal.SIGTERM, self._interrupt)


async def run_once(client: SessionClient) -> None:
    """Executes run_client_test once."""

    run_client_test(client)


async def run_continuously(client: SessionClient, handler: InterruptHandler) -> None:
    """Executes run_client_test until the process is interrupted."""

    while True:
        run_client_test(client)

        if handler.interrupted:
            break

        # now sleep for random duration between .5 to 1 second
        await asyncio.sleep(random.randrange(5, 11)/10)


def random_string(length: int = 12) -> str:
    """Generate a random string of fixed length."""

    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(length))


def run_client_test(client: SessionClient):
    """Initiate a Asherah Server session, encrypt sample data, decrypt the DRR, then compare
    the decrypted data to the original.
    """

    secret = f'my "secret" data - {random_string()}'.encode()

    logging.info('encrypting: %s', secret)
    drr = client.encrypt(secret)
    logging.info('received DRR')

    logging.info('decrypting DRR')
    data = client.decrypt(drr)
    logging.info('received decrypted data: %s', data)

    if secret != data:
        logging.info('oh no... something went terribly wrong!')
        return

    logging.info('test completed successfully')


async def _run_client(
        socket_file: string,
        partition_id: int,
        continuous: bool,
        handler: InterruptHandler
) -> None:
    partition = f'partitionid-{partition_id}'

    logging.info('starting session for %s', partition)
    with SessionClient(socket_file, partition) as client:
        if continuous:
            await run_continuously(client, handler)
        else:
            await run_once(client)


async def _main():
    parser = argparse.ArgumentParser(
        description='Connect to a local Asherah Server and execute a series of operations.')
    parser.add_argument(
        '--socket',
        default='/tmp/appencryption.sock',
        help='The unix domain socket the server is listening on.')
    parser.add_argument(
        '--num-clients',
        default=1,
        type=int,
        help='The total number of clients to run concurrently.')
    parser.add_argument(
        '--continuous',
        action='store_true',
        help='When present the client will run a continuous test.')
    args = parser.parse_args()

    handler = InterruptHandler()
    handler.start()

    logging.info('starting test')
    tasks = {asyncio.create_task(_run_client(args.socket, i, args.continuous, handler))
             for i in range(1, args.num_clients+1)}
    await asyncio.wait(tasks)


if __name__ == '__main__':
    logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
    asyncio.run(_main())
